### DISCUSSION TITLE
installing apt dependencies in plugin for docker

### DISCUSSION THREAD

**Discussion by wolflu05** [inventree/InvenTree#3562] [Q&A]
Hello,

I want to develop a CUPS label printing plugin for my printer. I already found a python library [`pycups`](https://pypi.org/project/pycups/) [[GitHub](https://github.com/OpenPrinting/pycups)] which works well for printing in my first tests.

The only issue I have is, that the following apt packages need to be installed inorder that the wheel building of the `pycups` packae can succeed. `gcc`, `libcups2-dev`, `python3-dev`

For a bare metal install I can document running `sudo apt install gcc libcups2-dev python3-dev`, but for docker this won't work or would require an execution after each recreate. Does anybody has a good idea how to solve this issue without maintaining an own inventree image which has these dependencies?

**Comment by matmair** [comment]
@wolflu05 I would leave that to the user. I have a mixin that runs user-level installs but that is not production ready.
There are several ways to achieve an install in docker - I probably would use a custom entrypoint.

**Reply by wolflu05** [reply]
@matmair Thank you for your idea via an entry point. But sorry to say that, I'm not really confident with that solution, but the problem is, I can't think of a better one. If I got this right, the wheel should only be build inside the worker right? Or is a plugin always build in both containers even if it doesn't need to be?

**Reply by matmair** [reply]
@wolflu05 not sure what you mean with both containers - we only provide one.

LMK if you come up with  a better solution, there are other ways to do this but it only goes downhill from here.

**Reply by wolflu05** [reply]
> not sure what you mean with both containers - we only provide one.

@matmair The production decker compose file provides two containers, the server and the worker container which use the same image. But if I do the trick with the entrypoints, do both containers need the custom entrypoints so that building the wheels will success for Pycups?


> LMK if you come up with a better solution, there are other ways to do this but it only goes downhill from here.

Sure I keep you posted. One thing I thought but have no idea if it will work is to search for some kind of prebuild hook of the wheels.

**Reply by matmair** [reply]
@wolflu05 Label printing is executed in the workers, if any are available, or in the server if not. You might need to install accordingly.



> > not sure what you mean with both containers - we only provide one.
> 
> @matmair The production decker compose file provides two containers, the server and the worker container which use the same image. But if I do the trick with the entrypoints, do both containers need the custom entrypoints so that building the wheels will success for Pycups?

> Sure I keep you posted. One thing I thought but have no idea if it will work is to search for some kind of prebuild hook of the wheels.

Wheels are [precompiled](https://realpython.com/python-wheels/#python-packaging-made-better-an-intro-to-python-wheels) - so no hooks - but you could use a [sdist](https://packaging.python.org/glossary/#term-source-distribution-or-sdist). This means you need to build a matching building pipeline.

**Reply by wolflu05** [reply]
Sorry I was using the wrong words. I meant a before sdist build hook. [This](https://stackoverflow.com/a/8828855) looks promising. My idea was to try installing the apt dependencies here if permissions are available, they are not already installed and OS uses apt for system dependencies.

**Reply by matmair** [reply]
sdists even can call native code so it should work for this.
Installing procedures for your dependencies might differ between the different plattforms. My suggestion would be to add a check early on if the executing platform is supported. That is often a problem with sdists.

**Reply by matmair** [reply]
@wolflu05 could you please select mark of the answers as the answer so that this stops showing up as unanswered?

**Reply by wolflu05** [reply]
@matmair Im not really sure what the answer to this problem is? Do you want me to create an issue for this contrib docker image tag?

**Reply by matmair** [reply]
@wolflu05 if you feel that is the right solution yes. I think sdist is an easier (and better scaleable) solution but that probably needs to be decided on a per-case basis.

**Reply by matmair** [reply]
I mainly want to clean out issues/get problems into FRs or bug reports before 0.9.0

**Reply by wolflu05** [reply]
> @wolflu05 if you feel that is the right solution yes. I think sdist is an easier (and better scaleable) solution but that probably needs to be decided on a per-case basis.

@matmair What would you think in this case for my cups plug-in?

**Reply by matmair** [reply]
@wolflu05 I think cups should work well with sdist

**Comment by SchrodingersGat** [comment]
There are a few options I can imagine:

### Custom Entrypoint

As @matmair says above you could have a custom entrypoint command which installs `libcups2-dev` every time you launch the container. Not super efficient.

### Custom Image

You could have an image which inherits from the `inventree/inventree` image and add a command to install cups when you build the image.

### External Container

You could have another docker container which has cups support and write some code to send label data to that. Not ideal, you'd have to write *two* pieces of code and it would be a tricky thing to get right.

### Install cups in the InvenTree image

Having cups support might be a good thing in general for InvenTree for label printing?

**Reply by matmair** [reply]
The image is already big, I would caution against a route of adding more stuff to it. Maybe creating another stage to the build that creates a `-contrib` image would be a good middle-ground?

**Reply by wolflu05** [reply]
That sounds nice, but doesn't generally solves this problem.

**Reply by matmair** [reply]
What problem is it not solving? It would enable you and anbody else who needs dependencies to get them in an image.
It should just not be the standard image as it would probably get pretty big.

**Reply by wolflu05** [reply]
oh ok, so this image would contain some more libs as they getting requested. I had a wrong view of it.
